[SVG] Handle `repeatEvent` for svg animations

Ahmad-S792 opened the pull request:
#### d7366f00065ab8ce5d0fd04f946c4225a5470149
<pre>
[SVG] Handle `repeatEvent` for svg animations
<a href="https://bugs.webkit.org/show_bug.cgi?id=275293">https://bugs.webkit.org/show_bug.cgi?id=275293</a>
<a href="https://rdar.apple.com/problem/129919749">rdar://problem/129919749</a>

Reviewed by Darin Adler.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://chromium.googlesource.com/chromium/blink/+/f7d3a4aaa8598d79b137dde63e9b401ea351e1e9">https://chromium.googlesource.com/chromium/blink/+/f7d3a4aaa8598d79b137dde63e9b401ea351e1e9</a>

This implements &apos;repeatEvent&apos; as per web specification [1]:

[1] <a href="https://www.w3.org/TR/SMIL3/smil-timing.html#q135">https://www.w3.org/TR/SMIL3/smil-timing.html#q135</a>

* Source/WebCore/dom/EventNames.json:
* Source/WebCore/svg/SVGAnimationElement.idl:
* Source/WebCore/svg/svgattrs.in:
* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(WebCore::SVGSMILElement::attributeChanged):
(WebCore::SVGSMILElement::progress):
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/repeat-event-expected.txt: Progressions
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/onrepeat-expected.txt: Ditto
* LayoutTests/imported/w3c/web-platform-tests/svg/idlharness.window-expected.txt: Ditto
* LayoutTests/platform/win/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-event-handler-content-attributes.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/set-event-handlers-content-attributes.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-event-handler-content-attributes.tentative-expected.txt:

Canonical link: <a href="https://commits.webkit.org/298715@main">https://commits.webkit.org/298715@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/16bd56817be33bc165988c7226d20419ff57215d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115634 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35297 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25820 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121681 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66150 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c6f3490d-7098-481c-a138-131daee6880e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117523 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35978 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43880 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87839 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/42469 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ece37666-4b00-4cba-97c5-f79a4c7f8f05) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118582 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28701 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103777 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68239 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27838 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21893 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65359 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98078 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22014 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124830 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42529 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31885 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96596 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42896 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99967 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96384 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24664 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41638 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19496 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38404 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42420 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47992 "Built successfully") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41893 "Built successfully") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45224 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43601 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/svg/animations/seeking-events-5-expected.txt:
```diff
@@ -1,3 +1,3 @@
 
-FAIL Seeking forwards does not dispatch 'repeatEvent' assert_true: Not expecting event, but got beginEvent event expected true got false
+FAIL Seeking forwards does not dispatch 'repeatEvent' assert_true: Not expecting event, but got repeatEvent event expected true got false
```
Does this mean we are dispatching events but we should not be? Is that something we can do right when adding `repeatEvent`, or does it need to be fixed separately?

Ahmad-S792 replied:
Yes - we need to fix the cases for dispatch events and in separate. We still don't expose 'onbegin' and others through IDL interface.

brentfulgham replied:
Could you file a Bugzilla (if it doesn't exist) for the lack of `beginEvent` (and it seems `endEvent`)? Then we can relate it to this issue, and we can mention that resolving those tasks will fix these test cases.

Ahmad-S792 replied:
With new changes - to not issues repeat and handle all repeat logic in `seekToTime` block, this test does not show any movement.

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/svg/animations/seeking-events-6-expected.txt:
```diff
@@ -1,3 +1,3 @@
 
-FAIL Seeking backwards does not dispatch 'repeatEvent' assert_true: Not expecting event, but got beginEvent event expected true got false
+FAIL Seeking backwards does not dispatch 'repeatEvent' assert_equals: Expected beginEvent event, but got repeatEvent event instead expected "beginEvent" but got "repeatEvent"
```
Does this mean we are dispatching events but we should not be? Is that something we can do right when adding `repeatEvent`, or does it need to be fixed separately?

brentfulgham replied:
@Ahmad-S792: Ditto -- file the bug about `beginEvent` and `endEvent`, and we can mention this test failure, too.

Ahmad-S792 replied:
With new changes - to not issues repeat and handle all repeat logic in `seekToTime` block, this test does not show any movement.

darinadler commented on Source/WebCore/svg/animation/SVGSMILElement.cpp:
```diff
@@ -1179,6 +1182,14 @@ bool SVGSMILElement::progress(SMILTime elapsed, SVGSMILElement& firstAnimation,
             smilEventSender().dispatchEventSoon(*this, eventNames().endEventEvent);
+
+        if (repeat) {
+            for (unsigned repeatEventCount = 1; repeatEventCount < repeat; repeatEventCount++)
```
The local variable name here is strange. It’s named with the word "count" but it’s not a count; it’s a loop index. The value `repeat` is actually the count. I think it would be better otherwise just use a traditional `i` or something like that for the loop index.

The loop here is strangely written. If `repeat` is 3, this will emit dispatch 2 repeat events, for values 1 and 2. Starting with 1 means this loop dispatches `repeat - 1` events. If this is intentional, then we need a comment explaining it; perhaps it’s supposed to dispatch `repeat - 1` events, and the dispatch outside the loop dispatches the last one? Either way, we also need a test that involves multiple repeat events, proving that it’s correct. I’m not sure we have one.

A more traditional way to write the loop would be:
```
for (unsigned i = 0; I < repeat - 1; ++i)
```
But we’d still need a comment explaining why we are doing one less than `repeat`.

Ahmad-S792 replied:
Yes - repeat(n) would be next step to fix. I just merged first part and then next would be repeat(n) case. :-)

We have repeatn cases in WPT and we fail them currently. :-(

brentfulgham replied:
It also seems like we potentially repeat in the `if (animationIsContributing) {` code path (prior to this), so we might have repeated one time already. That might be why this does one fewer than the repeat count, but that would only be true if `animationIsContributing` is true, and that's not checked here.

Then (as Darin points out) we potentially repeat again if `m_activeState == Inactive`, so it seems like all of this code would tend to repeat more times than expected.

Ahmad-S792 replied:
Updated and added comments to explained more clearly on how code is working. I looked into `repeat(n)` event bit as well and it seems, it is not specified as well - will check with Karl whether it is on track to specified and whether other browsers have plan to remove it (if they support it as of right now).

darinadler reviewed (changes requested):
I know we are merging code from Chromium, but one part of this code is unnecessarily mysterious and needs some better comments or possibly even to be changed.

darinadler commented on Source/WebCore/svg/SVGAnimationElement.idl:
```diff
@@ -39,6 +39,8 @@
     undefined endElement();
     undefined endElementAt(float offset);
+
+    attribute EventHandler onrepeat;
```
If the event bubbles then we might want this attribute on other elements too, any that could be an ancestor of this element; in other cases we put these attributes on all elements rather than on just the ones that emit the events. I'm surprised that we are adding it specifically to this element type only. I'm not saying it's wrong: we want to match other browsers' behavior.

brentfulgham commented on LayoutTests/imported/w3c/web-platform-tests/svg/animations/seeking-events-7-expected.txt:
Same comment applies here.

brentfulgham commented on Source/WebCore/svg/animation/SVGSMILElement.cpp:
```diff
@@ -1177,6 +1183,16 @@ bool SVGSMILElement::progress(SMILTime elapsed, SVGSMILElement& firstAnimation,
         if (m_activeState == Inactive || m_activeState == Frozen)
             smilEventSender().dispatchEventSoon(*this, eventNames().endEventEvent);
+
+        // FIXME: Add `repeat(n)` support where n would iterations. Below is named tentatively
```
When we add a `FIXME()` comment, we should file a Bugzilla documenting the action to take, and write this as `FIXME(<bugzilla ID>)`.

Ahmad-S792 replied:
Removed this comment - since repeat(n) is historical and not specified - will check with Karl to see what is the plan of it.

brentfulgham commented on Source/WebCore/svg/animation/SVGSMILElement.cpp:
```diff
@@ -1160,6 +1163,9 @@ bool SVGSMILElement::progress(SMILTime elapsed, SVGSMILElement& firstAnimation,
             startedActiveInterval();
 
+        if (repeat && repeat != m_lastRepeat)
+            smilEventSender().dispatchEventSoon(*this, eventNames().repeatEventEvent);
```
If we dispatched one of the `repeat` events here, maybe we should decrement `repeat` by 1 so we don't repeat an extra time?

Ahmad-S792 replied:
Done that by making sure that it is not '!seekToTime' to ensure that we don't dispatch incorrectly an extra time.

brentfulgham reviewed (commented):
Is it possible to land the parts of this that do not deal with `repeat(n)`, and simply do the right thing in the patch adding that logic?

Commit:
[SVG] Handle `repeatEvent` for svg animations
https://bugs.webkit.org/show_bug.cgi?id=275293
rdar://problem/129919749

Reviewed by Darin Adler.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: https://chromium.googlesource.com/chromium/blink/+/f7d3a4aaa8598d79b137dde63e9b401ea351e1e9

This implements 'repeatEvent' as per web specification [1]:

[1] https://www.w3.org/TR/SMIL3/smil-timing.html#q135

* Source/WebCore/dom/EventNames.json:
* Source/WebCore/svg/SVGAnimationElement.idl:
* Source/WebCore/svg/svgattrs.in:
* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(WebCore::SVGSMILElement::attributeChanged):
(WebCore::SVGSMILElement::progress):
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/repeat-event-expected.txt: Progressions
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/onrepeat-expected.txt: Ditto
* LayoutTests/imported/w3c/web-platform-tests/svg/idlharness.window-expected.txt: Ditto
* LayoutTests/platform/win/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-event-handler-content-attributes.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/set-event-handlers-content-attributes.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-event-handler-content-attributes.tentative-expected.txt:

Canonical link: https://commits.webkit.org/298715@main
